### PR TITLE
Step 11 - Split aggregator availability controller into local and remote part

### PR DESCRIPTION
**Pull request by sttts**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Towards kubernetes/enhancements#4080.

In a generic controlplane setup, the concept of service and/or endpoints might differ and is just not existent. This PR splits the availability controller kube-aggregator into a local and remote part. The later can be disabled or replaced in a generic controlplane. Local services (e.g. of native resources or CRDs) are still marked as available.

This PR does not change behaviour. It's pure code move and split of one controller handling both cases into two controllers handling either side.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller.go [hunk 1]
```diff
@@ -199,21 +195,12 @@ func (c *AvailableConditionController) sync(key string) error {
...
 		Status:             apiregistrationv1.ConditionTrue,
 		LastTransitionTime: metav1.Now(),
 	}
 
-	// local API services are always considered available
-	if apiService.Spec.Service == nil {
-		apiregistrationv1apihelper.SetAPIServiceCondition(apiService, apiregistrationv1apihelper.NewLocalAvailableAPIServiceCondition())
-		_, err := c.updateAPIServiceStatus(originalAPIService, apiService)
-		return err
-	}
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller.go:215]: @aojea this change drove me crazy. I am moving the if condition above the rest client creation. It failed integration tests. Then I moved it up line-by-line, eventually realizing that `restTransport, err := transport.New(transportConfig)` is the critical line. WTF. That was the moment I questions my career choices and the universe. Eventually, I read `TestAPIServerTransportMetrics` 😀

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/local/local_available_controller.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	originalAPIService, err := c.apiServiceLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		c.metrics.ForgetAPIService(key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if originalAPIService.Spec.Service != nil {
```
Comment (mjudeikis) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/local/local_available_controller.go:104]: Im reading these two `originalAPIService.Spec.Service  == nil` and `originalAPIService.Spec.Service  != nil` and wondering whether it is clear enough distinguishment? I think for now it ok, but in the future we might want to wrap this so shared helper maybe `apiservice.IsLocal()` or even psuedo type. I think few cycles on generic control plane will expose if this is enough :) but for now should be...

**Comment by mjudeikis** [comment]
/lgtm

I think it solves current generic control plane problems and helps us to iterate there to understand future needs and direction.

**Pull request merged**